Type Property For Button Block and its Migration Function

KirthikaKumar-K opened the pull request:
## Description
Type Property for Button Block and its Migration Function

## Changes Made
1. Added Type property to button block
2. Added Migration function.

## How to Test
Steps To Reproduce:
1. Create/Go to a to Drag and Drop app.
2. Drag in a button block.
3. Under block settings we would find the type property


## Video
https://github.com/user-attachments/assets/0d91f0a0-37da-494f-9178-3df2e74cf315

KirthikaKumar-K committed:
feat(renderer): button block changes

QodoAI-Agent commented:
## Title

Type Property For Button Block and its Migration Function

___
### **User description**
## Description
<!-- Provide a brief description of the changes in this PR -->

## Changes Made
<!-- List the key changes made in this PR -->

## How to Test
<!-- Explain how reviewers can test your changes -->
1. Steps to reproduce/test the behavior
2. Expected outcomes

## Notes
<!-- Any further notes regarding changes -->


___

### **PR Type**
Enhancement


___

### **Description**
- Add `type` attribute to ButtonBlockDef

- Expose `type` in ButtonBlock component

- Add UI setting for button `type`

- Implement state migration for `type`


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ButtonBlock.tsx</strong><dd><code>Enhance ButtonBlock with `type` property</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

libs/renderer/src/components/block-defaults/button-block/ButtonBlock.tsx

<li>Include <code>type</code> in ButtonBlockDef data<br> <li> Pass <code>type={data?.type}</code> to StyledButton<br> <li> Add console.log for debugging


</details>


  </td>
  <td><a href="https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-04b4f2cf4edfce72877451191e7ca64317415feae3d21b1e85e381d693c68494">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>migrate__1_0_0_alpha_11__to___1_0_0_alpha_12.ts</strong><dd><code>Migration: default button `type` assignment</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

libs/renderer/src/store/state/migration/migrate__1_0_0_alpha_11__to___1_0_0_alpha_12.ts

<li>Add migration for default button <code>type</code><br> <li> Iterate blocks and set missing <code>type</code><br> <li> Return updated state with defaults


</details>


  </td>
  <td><a href="https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-50e7093e026af42bd176c721955ac7b85f336dd85223c304a3612cec2517fa1f">+33/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>config.tsx</strong><dd><code>Configure button `type` UI setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

libs/renderer/src/components/block-defaults/button-block/config.tsx

<li>Initialize <code>type</code> in default button data<br> <li> Add SelectInputSettings for <code>type</code> property


</details>


  </td>
  <td><a href="https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-1bdcf5e41a98730b90f6d2980c004f07feea8b198df686bb5b29aa5dc0238130">+25/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>MigrationManager.ts</strong><dd><code>Bump state version and register migration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

libs/renderer/src/store/state/migration/MigrationManager.ts

<li>Import new migration module alpha.11→alpha.12<br> <li> Update STATE_VERSION to <code>1.0.0-alpha.12</code><br> <li> Register migration in migrations map


</details>


  </td>
  <td><a href="https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-5a273187f5383e1f46eb36c3aa51b8042a2ed2c690649a5a43c8f4cdf40a01e2">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>default-menu.ts</strong><dd><code>Default menu: add button `type`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/client/src/components/blocks-workspace/menus/default-menu.ts

- Set default button `type` to `button`


</details>


  </td>
  <td><a href="https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-1a960304ff5e3fb3c4eca077f03c1c800d2bb038d407cf962ac3567ba45a869c">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

QodoAI-Agent commented:
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-04b4f2cf4edfce72877451191e7ca64317415feae3d21b1e85e381d693c68494R67-R67'><strong>Debug Logging</strong></a>

A console.log statement remains in production code, which can clutter logs and reveal internal data. Remove or guard this debug output before merging.
</summary>

```typescript
console.log("ButtonBlock", data);
```

</details>

<details><summary><a href='https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-50e7093e026af42bd176c721955ac7b85f336dd85223c304a3612cec2517fa1fR12-R26'><strong>Migration Side Effects</strong></a>

The migration directly mutates the input state and uses `hasOwnProperty` unsafely. Consider cloning objects or using a safer pattern to avoid unintended state mutations and prototype pollution.
</summary>

```typescript
for (const key in blocks) {
    // Check if the current property is a direct property of blocks
    if (blocks.hasOwnProperty(key)) {
        const block = blocks[key]; // Get the block associated with the current key

        // Check if the block is a button widget
        if (block.widget === "button") {
            // Set the type to "button" if it's not already defined in block's data
            if (!block.data.type) {
                block.data.type = "button";
            }
        }
        blocks[key] = block;
    }
}
```

</details>

</td></tr>
</table>

QodoAI-Agent commented:
## PR Code Suggestions ✨
<!-- ab4b37b -->

Latest suggestions up to ab4b37b

<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=1>Possible issue</td>
<td>



<details><summary>Provide default button type</summary>

___


**Ensure the <code>type</code> prop is never undefined by providing a fallback of <code>"button"</code>. This <br>prevents unintended default form submissions when <code>data.type</code> is missing.**

[libs/renderer/src/components/block-defaults/button-block/ButtonBlock.tsx [75]](https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-04b4f2cf4edfce72877451191e7ca64317415feae3d21b1e85e381d693c68494R75-R75)

```diff
-type={data?.type}
+type={data.type || 'button'}
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: The fallback `type={data.type || 'button'}` prevents unintended form submissions when `data.type` is undefined, improving reliability with moderate impact.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

___

#### Previous suggestions


<details><summary>Suggestions up to commit 41423b7</summary>
<br><table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Guard migration on valid data</summary>

___


**Add a guard to skip entries without a valid <code>data</code> object to prevent runtime errors on <br>malformed or unexpected block entries. This ensures only blocks with a <code>data</code> property <br>get processed.**

[libs/renderer/src/store/state/migration/migrate__1_0_0_alpha_11__to___1_0_0_alpha_12.ts [12-18]](https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-50e7093e026af42bd176c721955ac7b85f336dd85223c304a3612cec2517fa1fR12-R18)

```diff
 for (const [key, block] of Object.entries(blocks)) {
+    if (!block?.data || typeof block.data !== 'object') {
+        newBlocks[key] = block;
+        continue;
+    }
     // Clone the block to avoid mutation
     const newBlock = { ...block, data: { ...block.data } };
     if (newBlock.widget === "button" && !newBlock.data.type) {
         newBlock.data.type = "button";
     }
     newBlocks[key] = newBlock;
 }
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Adding a check for `block?.data` prevents runtime errors on malformed entries, making the migration more robust without breaking existing logic.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Add default button type</summary>

___


**Provide a default fallback for the button type to avoid unintended form submissions <br>when <code>data.type</code> is undefined. This ensures the button never defaults to HTML's <br>implicit "submit" type.**

[libs/renderer/src/components/block-defaults/button-block/ButtonBlock.tsx [75]](https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-04b4f2cf4edfce72877451191e7ca64317415feae3d21b1e85e381d693c68494R75-R75)

```diff
-type={data?.type}
+type={data?.type ?? 'button'}
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: The fallback `data?.type ?? 'button'` prevents unintended form submissions when `data.type` is undefined, improving component safety with minimal impact.


</details></details></td><td align=center>Low

</td></tr><tr><td rowspan=1>General</td>
<td>



<details><summary>Reorder dropdown options</summary>

___


**Reorder the type options to list "button" first, matching the default data and <br>reducing confusion in the dropdown.**

[libs/renderer/src/components/block-defaults/button-block/config.tsx [77-90]](https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-1bdcf5e41a98730b90f6d2980c004f07feea8b198df686bb5b29aa5dc0238130R77-R90)

```diff
 options={[
+    {
+        value: "button",
+        display: "Button",
+    },
     {
         value: "submit",
         display: "Submit",
-    },
-    {
-        value: "button",
-        display: "Button",
     },
     {
         value: "reset",
         display: "Reset",
     },
 ]}
```
<details><summary>Suggestion importance[1-10]: 3</summary>

__

Why: Reordering options to match the default reduces minor confusion for users but does not affect functionality.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

</details>
<details><summary>Suggestions up to commit e3390b8</summary>
<br><table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Add default button type</summary>

___


**Provide an explicit fallback to <code>"button"</code> to avoid unintended <code>"submit"</code> behavior when <br><code>data.type</code> is undefined. This ensures consistent default action.**

[libs/renderer/src/components/block-defaults/button-block/ButtonBlock.tsx [76]](https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-04b4f2cf4edfce72877451191e7ca64317415feae3d21b1e85e381d693c68494R76-R76)

```diff
-type={data?.type}
+type={data.type ?? "button"}
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Providing a fallback of `"button"` ensures the HTML button does not default to `"submit"` unexpectedly for undefined `data.type`.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Guard missing block data</summary>

___


**Guard against missing <code>data</code> to avoid runtime exceptions. Ensure <code>block.data</code> exists <br>before reading or writing the <code>type</code> property.**

[libs/renderer/src/store/state/migration/migrate__1_0_0_alpha_11__to___1_0_0_alpha_12.ts [20-22]](https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-50e7093e026af42bd176c721955ac7b85f336dd85223c304a3612cec2517fa1fR20-R22)

```diff
-if (!block.data.type) {
-    block.data.type = "button";
+if (block.data) {
+    if (!block.data.type) {
+        block.data.type = "button";
+    }
 }
```
<details><summary>Suggestion importance[1-10]: 4</summary>

__

Why: Adding a check for `block.data` prevents potential runtime errors but is a minor defensive improvement.


</details></details></td><td align=center>Low

</td></tr><tr><td rowspan=1>General</td>
<td>



<details><summary>Remove debug logging</summary>

___


**Remove or guard this debug logging to prevent noisy output in production. If you <br>need runtime diagnostics, consider using a debug logger or conditional log based on <br>environment.**

[libs/renderer/src/components/block-defaults/button-block/ButtonBlock.tsx [67]](https://github.com/SEMOSS/semoss-ui/pull/1387/files#diff-04b4f2cf4edfce72877451191e7ca64317415feae3d21b1e85e381d693c68494R67-R67)

```diff
-console.log("ButtonBlock", data);
+// Remove debug log or replace with a debug-level logger if needed
```
<details><summary>Suggestion importance[1-10]: 5</summary>

__

Why: Removing the stray `console.log` prevents noisy output in production and improves code cleanliness.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

</details>

KirthikaKumar-K committed:
feat(renderer): fixed PR reviews

KirthikaKumar-K committed:
Merge remote-tracking branch 'origin/dev' into 1357_addtypeproperty_buttonblock

KirthikaKumar-K committed:
Merge remote-tracking branch 'origin/dev' into 1357_addtypeproperty_buttonblock

KirthikaKumar-K committed:
Merge remote-tracking branch 'origin/dev' into 1357_addtypeproperty_buttonblock

j-adam-connor commented:
Hello! Just a couple of things I'd point out before beginning review for this:
- The PR description is incomplete.
- The PR agent identifies some potential issues that may need your attention [here](https://github.com/SEMOSS/semoss-ui/pull/1387#issuecomment-3007392045)

Thank you!

KirthikaKumar-K committed:
Merge remote-tracking branch 'origin/dev' into 1357_addtypeproperty_buttonblock

KirthikaKumar-K commented:
> Hello! Just a couple of things I'd point out before beginning review for this:
> 
> * The PR description is incomplete.
> * The PR agent identifies some potential issues that may need your attention [here](https://github.com/SEMOSS/semoss-ui/pull/1387#issuecomment-3007392045)
> 
> Thank you!

Hi @j-adam-connor ,
- Added a proper description.
- Fixed the potential issues identified by the PR agent and committed the changes in one the previous commits. Attaching it here for easy access [https://github.com/SEMOSS/semoss-ui/pull/1387/commits/bc997ff3390f1ad41cc520ef71c7cb188216e502](url)

Thank you !

j-adam-connor commented on libs/renderer/src/store/state/migration/migrate__1_0_0_alpha_11__to___1_0_0_alpha_12.ts:
```diff
@@ -0,0 +1,26 @@
+
+    // Return a new state object with updated blocks
+    return { ...state, blocks: newBlocks };
+}
```
This curly brace should be indented with a trailing comma. Double check your formatter is configured correctly.

j-adam-connor commented on libs/renderer/src/store/state/migration/migrate__1_0_0_alpha_11__to___1_0_0_alpha_12.ts:
It looks like we have an established pattern that's a bit different from this one where we destructure the state into `newState`; e.g., `const newState = { ... state };` and then we ultimately return `newState`

KirthikaKumar-K replied:
Hi @j-adam-connor, I've made the updates based on your comments. You can review the changes [here](https://github.com/SEMOSS/semoss-ui/pull/1387/commits/0303a0a820b4db80052956a5c78ffa0c95fd8d61)

KirthikaKumar-K committed:
Merge remote-tracking branch 'origin/dev' into 1357_addtypeproperty_buttonblock

KirthikaKumar-K committed:
feat(renderer): fixed the pr comments

KirthikaKumar-K committed:
Merge remote-tracking branch 'origin/dev' into 1357_addtypeproperty_buttonblock

j-adam-connor commented:
Hey, looks like there are some conflicts. Please resolve, and we'll get this merged. Thanks!

KirthikaKumar-K committed:
Merge remote-tracking branch 'origin/dev' into 1357_addtypeproperty_buttonblock

QodoAI-Agent commented:
**Changelog updates:** 🔄

## 2025-07-17 [*](https://github.com/SEMOSS/semoss-ui/pull/1387)

### Added
- Support for a new `type` property on button blocks.
- Migration to backfill existing button blocks with a default type.


>to commit the new content to the CHANGELOG.md file, please type:
>'/update_changelog --pr_update_changelog.push_changelog_changes=true'
